Add __future__.annotations in forward refs section (#9390)
python/mypy: docs/source/kinds_of_types.rst

   class A:
       ...

> In cases when a type hint contains names that have not been defined yet,

Reviewer: I think the previous wording is fine to keep.

Author: Okay, I'll keep previous

   class A:
       ...

> Starting from Python 3.7 (:pep:`563`) using of string literal type is no longer required,
> following special import is enough:

Reviewer: ```suggestion
Starting from Python 3.7 (:pep:`563`), you can add the special import ``from __future__ import annotations``,
which makes the use of string literals unnecessary:
```

Author: Thank you, that one is better

> which makes the use of string literals unnecessary:

Reviewer 2: ```suggestion
which makes the use of string literals in annotations unnecessary:
```
Just so it contradicts the note a little less

Author: Yes, this looks better, thanks

.. code-block:: python

   from __future__ import annotations

   def f(x: A) -> None:  # OK
       ...

   class A:
       ...

.. note::

>     Even with the ``__future__`` import, there are some scenarios that still

Reviewer 2: ```suggestion
    Even with the ``__future__`` import, there are some scenarios that could still
```

Author: +

>     Despite adding special import, there are some scenarios that still
>     require working around forward references with string literals:

Reviewer: ```suggestion
    Even with the ``__future__`` import, there are some scenarios that still
    require string literals:
```

Author: Thank you for the improvements

>     require string literals:

Reviewer 2: ```suggestion
    require string literals, typically involving use of forward references or generics in:
```

Author: Thanks, that's more accurate

    * :ref:`type aliases <type-aliases>`;
    * :ref:`casts <casts>`;
    * type definitions (see :py:class:`~typing.TypeVar`, :py:func:`~typing.NewType`, :py:class:`~typing.NamedTuple`);
    * base classes.

    .. code-block:: python

        # base class example

>         class A(Tuple['B', 'C']): ...# OK

Reviewer: ```suggestion
        class A(Tuple['B', 'C']): ...  # OK
```

Author: Oh, my bad

   class A: pass

> String literal types are never needed in `# type:` comments.

Reviewer: Or in stub files (not directly related to the topic of this PR, but we might as well add it).

Author: Mentioned stub files

> String literal types are never needed in `# type:` comments and :ref:`stub files <stub-files>`.

Author: This change was [proposed](https://github.com/python/mypy/pull/9390#discussion_r481259039) by @JelleZijlstra. I used to use double backticks for code, so I wasn't sure about highlighting of "code-comments", so I removed double one. I will commit changes as soon as you and @JelleZijlstra will reach an aggreement about backticks.

> String literal types are never needed in `# type:` comments and :ref:`stub files <stub-files>`.

Reviewer 2: ```suggestion
String literal types are never needed in ``# type:`` comments and :ref:`stub files <stub-files>`.
```

> String literal types are never needed in `# type:` comments and :ref:`stub files <stub-files>`.

Reviewer 2: We use double backticks for, eg ``# type: ignore`` in several other places, so I'm going to go ahead and make the change.
